File: web/src/main/java/be/ucll/runetracker/database/DatabaseService.java
package be.ucll.runetracker.database;

import be.ucll.runetracker.domain.DataPoint;
import be.ucll.runetracker.domain.RunescapeUser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collection;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class DatabaseService {
    private static final Logger logger = LoggerFactory.getLogger(DatabaseService.class);

    private final RunescapeUserDatabase runescapeUserDatabase;
    private final DataPointDatabase dataPointDatabase;

    public DatabaseService(RunescapeUserDatabase runescapeUserDatabase, DataPointDatabase dataPointDatabase) {
        this.runescapeUserDatabase = runescapeUserDatabase;
        this.dataPointDatabase = dataPointDatabase;
    }

    // USERS
    public Optional<RunescapeUser> getUser(int id) {
        logger.info("Getting user with id: " + id);
        Optional<RunescapeUser> user = runescapeUserDatabase.get(id);
        if (user.isPresent()) {
            String debugString = datapointsToString(user.get().getDataPoints());
            logger.info(debugString);
        } else {
            logger.info("Could not find user with id: " + id);
        }

        return user;
    }

    private String datapointsToString(Collection<DataPoint> dataPoints) {
        logger.info("" + dataPoints.size());

        if (dataPoints.isEmpty()) {
            return "No datapoints available";
        }

        return dataPoints
                .stream()
                .map(DataPoint::toString)
                .collect(Collectors.joining("\n", "Datapoints: \n", ""));
    }

    public Stream<RunescapeUser> getAllUsers() {
        logger.debug("Getting all users");
        return runescapeUserDatabase.all();
    }

    public void addUser(RunescapeUser user) {
        logger.info("Adding user");
        runescapeUserDatabase.add(user);
    }

    public void deleteUser(RunescapeUser user) {
        logger.info("Deleting user with id: " + user.getId());
        runescapeUserDatabase.delete(user);
    }

    public void updateUser(RunescapeUser user) {
        logger.info("Updating user with id: " + user.getId());
        runescapeUserDatabase.update(user);
    }

    // DATAPOINTS
    public Optional<DataPoint> getDatapoint(int id) {
        logger.debug("Getting datapoint with id: " + id);
        return dataPointDatabase.get(id);
    }

    public Stream<DataPoint> getAllDatapoints() {
        logger.debug("Getting all datapoints");
        return dataPointDatabase.all();
    }

    public void addDatapoint(DataPoint dataPoint) {
        logger.info("Adding datapoint");
        dataPointDatabase.add(dataPoint);
    }

    public void deleteDatapoint(DataPoint dataPoint) {
        logger.info("Deleting datapoint with id: " + dataPoint.getId());
        dataPointDatabase.delete(dataPoint);
    }
}

File: web/src/main/java/me/robinvdb/runetracker/database/DatabaseService.kt
package me.robinvdb.runetracker.database

import me.robinvdb.runetracker.domain.DataPoint
import me.robinvdb.runetracker.domain.RunescapeUser
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.Optional
import java.util.stream.Collectors
import java.util.stream.Stream

class DatabaseService(private val runescapeUserDatabase: RunescapeUserDatabase, private val dataPointDatabase: DataPointDatabase) {

    val allUsers: Stream<RunescapeUser>
        get() {
            logger.debug("Getting all users")
            return runescapeUserDatabase.all()
        }

    val allDatapoints: Stream<DataPoint>
        get() {
            logger.debug("Getting all datapoints")
            return dataPointDatabase.all()
        }

    // USERS
    fun getUser(id: Int): Optional<RunescapeUser> {
        logger.info("Getting user with id: $id")
        val user = runescapeUserDatabase.get(id)
        if (user.isPresent) {
            val debugString = datapointsToString(user.get().dataPoints)
            logger.info(debugString)
        } else {
            logger.info("Could not find user with id: $id")
        }

        return user
    }

    private fun datapointsToString(dataPoints: Collection<DataPoint>): String {
        logger.info(dataPoints.size.toString(10))

        if (dataPoints.isEmpty()) {
            return "No datapoints available"
        } else {
            return dataPoints.joinToString("\n", "Datapoints: \n", "") { it.toString() }
        }
    }

    fun addUser(user: RunescapeUser) {
        logger.info("Adding user")
        runescapeUserDatabase.add(user)
    }

    fun deleteUser(user: RunescapeUser) {
        logger.info("Deleting user with id: " + user.id!!)
        runescapeUserDatabase.delete(user)
    }

    fun updateUser(user: RunescapeUser) {
        logger.info("Updating user with id: " + user.id!!)
        runescapeUserDatabase.update(user)
    }

    // DATAPOINTS
    fun getDatapoint(id: Int): Optional<DataPoint> {
        logger.debug("Getting datapoint with id: $id")
        return dataPointDatabase.get(id)
    }

    fun addDatapoint(dataPoint: DataPoint) {
        logger.info("Adding datapoint")
        dataPointDatabase.add(dataPoint)
        val user = dataPoint.user
        user!!.addDataPoint(dataPoint)
        runescapeUserDatabase.update(user)
    }

    fun deleteDatapoint(dataPoint: DataPoint) {
        logger.info("Deleting datapoint with id: " + dataPoint.id!!)
        dataPointDatabase.delete(dataPoint)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DatabaseService::class.java)
    }
}
